docs/denotational-design.md

```diff
@@ -1,214 +1,260 @@
+      IsRepo               -> []
+
+fileName :: File -> Label
+μ fileName file = fst (μ file)
 
--- Getting the [Directory] from CommitHistory should be equivalent to building
--- the [File] and getting their Directory
-files_and_directories :: Property
-files_and_directories = ∀ history. getHistoryDirectories history ≡ map fileDirectory (getFiles (getCommits history))
+findFile :: NonEmpty Label -> Directory -> Maybe File
+μ findFile (label :| labels) (Directory (label', contents)) =
+  if label == label' then go labels contents else Nothing
+  where
+    findFileWithLabel :: Foldable f => Label -> f DirectoryContents -> Maybe File
+    findFileWithLabel label = find (\artefact -> case content of
+      File (fileLabel, _) -> fileLabel == label
+      Directory _         -> False
+      IsRepo              -> False)
+
+    go :: [Label] -> NonEmpty DirectoryContents -> Just File
+    go []             _        = Nothing
+    go [label]        contents = findMaybe (fileWithLabel label) contents
+    go (label:labels) contents = (go labels . snd) <$> find ((label ==) . fst) onlyDirectories contents
+
+onlyDirectories :: Foldable f => f DirectoryContents -> [Directory]
+μ onlyDirectories = fmapMaybe (\content -> case content of
+  d@(Directory _) -> Just d
+  File _          -> Nothing
+  IsRepo          -> Nothing) . toList
+
+getSubDirectories :: Directory -> [Directory]
+μ getSubDirectories directory = foldMap f $ snd (μ directory)
+  where
+    f :: DirectoryContents -> [Directory]
+    f = \case
+      d@(Directory _) -> [d]
+      File _          -> []
+      IsRepo          -> []
+
+-- Definition elided
+findDirectory :: NonEmpty Label -> Directory -> Maybe Directory
+
+-- Definition elided
+fuzzyFind :: Label -> [Directory]
+
+-- A Git Snapshot is grabbing the HEAD commit of your History
+-- and turning it into a Directory
+gitSnapshot :: Snapshot Commit
+μ gitSnapshot commits = second (\root -> root <> getDirectoryPtr $ Nel.head commits) emptyRepoRoot
+
+-- Opaque and defined by the backend
+getDirectoryPtr :: Commit -> Directory
+
+-- A Pijul history is semantically applying the patches in a
+-- topological order and achieving the Directory view.
+pijulHistory :: Snapshot Patch
+μ pijulHistory = foldl pijulMagic emptyRepoRoot
+
+-- Opaque and defined by the backend
+pijulMagic :: Patch -> Directory -> Directory
+
+-- Get the current History we are working with.
+getHistory :: Browser a (History a)
+μ getHistory = get
+
+setHistory :: History a -> Browser a ()
+μ setHistory = put
+
+-- Get the current Directory in the Browser
+getDirectory :: Browser a Directory
+μ getDirectory = do
+  hist <- get
+  applySnapshot <- ask
+  pure $ applySnapshot hist
+
+-- We modify the history by changing the internal history state.
+switchHistory :: ([a] -> [a]) -> Browser a b
+μ switchHistory f = modify f
+
+-- View the history up to a given point
+viewAt :: Eq a => a -> Browser a b
+μ viewAt a = switchHistory (dropWhile (/= a))
```

Comment:
Oh no... I missed this when changing `History` to `NonEmpty`. This doesn't quite work. Need something like `Maybe (Browser a b)` because we can't be working over an empty history if we can't find the artifact.

Reply:
But that type sig doesn't seem to make sense... maybe something like:
```haskell
type ViewResult
  = Fail
  | Succes

viewAt :: Eq a => a -> Browser a ViewResult
μ viewAt a = do
  hist <- get
  let new_hist = nonEmpty $ Nel.dropWhile (/= a) hist
  case new_hist of
    Nothing -> pure Failure
    Just h -> setHistory h *> pure Success
```

Reply:
@kim Any thoughts on making this nicer?

Reply:
Hm, yeah, it's not entirely satisfying, because the caller needs to scrutinise the monadic result, while the state stays the same. The honest thing to do would be to diverge (`throw`), or add an `ExceptT` layer.

Or, let the caller decide which `History` to default to:

```haskell
findSuffix :: Eq a => a -> History a -> Maybe (History a)
μ findSuffix a = nonEmpty . Nel.dropWhile (/= a)

viewAt :: (History a -> Maybe (History a)) -> History a -> Browser a b
μ viewAt f def = switchHistory (fromMaybe def . f)
```

Reply:
Ya that seems cleaner!